code/datums/status_effects/debuffs/fire_stacks.dm

```diff
@@ -0,0 +1,376 @@
+	stacks = max(0, min(stack_limit, new_stacks))
+	cache_stacks()
+
+/datum/status_effect/fire_handler/proc/adjust_stacks(new_stacks)
+	stacks = max(0, min(stack_limit, stacks + new_stacks))
+	cache_stacks()
+
+/**
+ * Refresher for mob's fire_stacks
+ */
+
+/datum/status_effect/fire_handler/proc/cache_stacks()
+	owner.fire_stacks = 0
+	var/was_on_fire = owner.on_fire
+	owner.on_fire = FALSE
+	for(var/datum/status_effect/fire_handler/possible_fire in owner.status_effects)
+		owner.fire_stacks += possible_fire.stacks * possible_fire.stack_modifier
+
+		if(!istype(possible_fire, /datum/status_effect/fire_handler/fire_stacks))
+			continue
+
+		var/datum/status_effect/fire_handler/fire_stacks/our_fire = possible_fire
+		if(our_fire.on_fire)
+			owner.on_fire = TRUE
+
+	if(was_on_fire && !owner.on_fire)
+		owner.clear_alert(ALERT_FIRE)
+	else if(!was_on_fire && owner.on_fire)
+		owner.throw_alert(ALERT_FIRE, /atom/movable/screen/alert/fire)
+
+/**
+ * Used to update owner's effect overlay
+ */
+
+/datum/status_effect/fire_handler/proc/update_overlay()
+
+/datum/status_effect/fire_handler/fire_stacks
+	id = "fire_stacks" //fire_stacks and wet_stacks should have different IDs or else has_status_effect won't work
+
+	enemy_types = list(/datum/status_effect/fire_handler/wet_stacks)
+	stack_modifier = 1
+
+	/// If we're on fire
+	var/on_fire = FALSE
+	/// A weakref to the mob light emitter
+	var/datum/weakref/firelight_ref
+	/// Type of mob light emitter we use when on fire
+	var/firelight_type = /obj/effect/dummy/lighting_obj/moblight/fire
+	/// Stores current fire overlay icon state, for optimisation purposes
+	var/last_icon_state
+
+/datum/status_effect/fire_handler/fire_stacks/tick(delta_time, times_fired)
+	if(stacks <= 0)
+		qdel(src)
+		return TRUE
+
+	if(!on_fire || isanimal(owner))
+		return TRUE
+
+	adjust_stacks(-0.05 * delta_time)
+
+	if(iscyborg(owner))
+		adjust_stacks(-0.5 * delta_time)
+
+	if(stacks <= 0)
+		qdel(src)
+		return TRUE
+
+	var/datum/gas_mixture/air = owner.loc.return_air()
+	if(!air.gases[/datum/gas/oxygen] || air.gases[/datum/gas/oxygen][MOLES] < 1)
+		qdel(src)
+		return TRUE
+
+	deal_damage(delta_time, times_fired)
+	update_overlay()
+
+/**
+ * Proc that handles damage dealing and all special effects
+ *
+ * Arguments:
+ * - delta_time
+ * - times_fired
+ *
+ */
+
+/datum/status_effect/fire_handler/proc/deal_damage(delta_time, times_fired)
+	if(ishuman(owner))
+		var/mob/living/carbon/human/victim = owner
+		SEND_SIGNAL(victim, COMSIG_HUMAN_BURNING)
+		burn_stuff(delta_time, times_fired)
+		var/no_protection = FALSE
+		if(victim.dna && victim.dna.species)
+			no_protection = victim.dna.species.handle_fire(victim, delta_time, times_fired, no_protection)
+		harm_human(delta_time, times_fired, no_protection)
+
+	if(isalien(owner))
+		owner.adjust_bodytemperature(BODYTEMP_HEATING_MAX * 0.5 * delta_time)
+
+	var/turf/location = get_turf(owner)
+	location.hotspot_expose(700, 25 * delta_time, TRUE)
```

Comment:
I'm not a massive fan of this pattern, If you're going to handle subtypes like this I think inheritance is a better way of doing it

Reply:
I still feel this way